src/imp/run.c

```diff
@@ -0,0 +1,333 @@
+     *   - 'path' must exist and be an absolute path
+     */
+    if (!(path = cf_string (cf_get_in (cf, "path")))
+        || path[0] == '\0'
+        || path[0] != '/')
+        imp_die (1, "run: %s: path is missing or invalid", name);
+
+    return cf;
+}
+
+static bool run_user_allowed (const cf_t *cf_run)
+{
+    struct passwd *pwd;
+
+    if (!(pwd = getpwuid (getuid ())) || !pwd->pw_name)
+        imp_die (1, "Unable to lookup user");
+
+    return cf_array_contains (cf_get_in (cf_run, "allowed-users"),
+                              pwd->pw_name);
+}
+
+static bool run_env_var_allowed (const cf_t *allowed_env,
+                                 const char *name)
+{
+    if (strcmp (name, "FLUX_JOB_ID") == 0
+        || strcmp (name, "FLUX_JOB_USERID") == 0)
+        return true;
+    return cf_array_contains_match (allowed_env, name);
+}
+
+/*  Return a kv structure with the environment for this run command.
+ */
+static struct kv *get_run_env (struct kv *kv, const cf_t *allowed_env)
+{
+    struct kv *kv_env;
+    const char *var = NULL;
+
+    if (!(kv_env = kv_split (kv, "IMP_RUN_ENV_")))
+        return NULL;
+
+    while ((var = kv_next (kv_env, var))) {
+        if (!run_env_var_allowed (allowed_env, var))
+            kv_delete (kv_env, var);
+    }
+
+    /*  Capture uid that ran the imp as the "owner" */
+    if (kv_put (kv_env,
+                "FLUX_OWNER_USERID",
+                KV_INT64,
+                (int64_t) getuid()) < 0)
+        imp_die (1, "failed to put FLUX_IMP_USERID in environment");
+
+    return kv_env;
+}
+
+static void __attribute__((noreturn))
+imp_run (const char *name,
+         const cf_t *run_cf,
+         struct kv *kv_env)
+{
+    const char *path;
+    struct passwd *pwd;
+    char **env;
+    const char *args[2];
+    int exit_code;
+
+    if (!(path = cf_string (cf_get_in (run_cf, "path")))
+        || path[0] != '/')
+        imp_die (1, "run: %s: invalid path", name);
+
+    if (path[0] != '/')
+        imp_die (1, "run %s: relative path not allowed", name);
+
+    /*  Get passwd entry for current user to set HOME and USER */
+    if (!(pwd = getpwuid (getuid ())))
+        imp_die (1, "run: failed to find target user");
+
+    /*  Set HOME and USER */
+    if (kv_put (kv_env, "HOME", KV_STRING, pwd->pw_dir) < 0
+        || kv_put (kv_env, "USER", KV_STRING, pwd->pw_name) < 0)
+        imp_die (1, "run: failed to set HOME and USER in environment");
+
+    /*  Set PATH to a sane default */
+    if (kv_put (kv_env,
+                "PATH",
+                KV_STRING,
+                "/usr/sbin:/usr/bin:/sbin:/bin") < 0)
+        imp_die (1, "failed to put default PATH in environment");
+
+    if (kv_expand_environ (kv_env, &env) < 0)
+        imp_die (1, "Unable to set %s environment", name);
+
+    if (chdir ("/") < 0)
+        imp_die (1, "run: failed to chdir to /");
+
+    args[0] = path;
+    args[1] = NULL;
+    execvpe (path, (char **) args, env);
```

Comment:
`execvpe()` searches for executable in PATH.  Since we're ensuring the path is not relative, would `execve()` be more appropriate?

Reply:
Yeah, not sure how I ended up with `execvpe()` there :shrug: It should definitely be `execve()`. Thanks for catching that.